### ISSUE TITLE
Automatic ordering of layout transitions

### ISSUE THREAD

**Issue by philiptaylor** [KhronosGroup/Vulkan-Docs#302]
https://www.khronos.org/registry/vulkan/specs/1.0/xhtml/vkspec.html#synchronization-image-memory-barrier says:

> Layout transitions that are performed via image memory barriers are automatically ordered against other layout transitions, including those that occur as part of a render pass instance.

I have no idea what this means. What was intended here?

The previous paragraph already requires memory dependencies (implying execution dependencies) between operations on the image before/after the transition, which I believe includes other transitions before/after the transition, so the application is already responsible for providing a sufficient ordering between those transitions. If the application violates that (e.g. putting two `VkImageMemoryBarrier` with the same `image` in a single `vkCmdPiplineBarrier`, or doing transitions on the same image in two separate queues), I don't see why it would be helpful or even possible to automatically define some execution-dependency total ordering between them, and I don't see what other kind of ordering would be relevant.

Maybe it's trying to refer to the way that automatic layout transitions in render passes are defined to create automatic dependencies in certain cases (but not all cases), or something like that, but it seems very unclear to me.

**Comment by krOoze** [comment]
I think, it simply describes a missing piece in Execution Order (in single queue).

There are three types of commands: Action, State and Synchronization.

State commands must be executed in order (well, only to the point the other commands can't detect any difference).

Action commands can be reordered and overlap, regulated by Synchronization commands (and Implicit rules). 

Synchronization commands can reorder between themselves (I think -- nothing bad should happen if you switch two Buffer Barriers that are next to each other, right?), with the exception of Image Barriers (user needs to be able to chain them, because of the image layout -- if I record change of layout A to B and then B to C, it would not work reordered -- hence "ordered automatically against themselves")

**Comment by philiptaylor** [comment]
> (I think -- nothing bad should happen if you swap two Buffer Barriers that are next to each other, right?)

(It will break things like:

    vkCmdPipelineBarrier(
        srcStageMask=FRAGMENT_SHADER,
        dstStageMask=FRAGMENT_SHADER,
        VkBufferMemoryBarrier { srcAccessMask=SHADER_WRITE, dstAccessMask=0 }
    );
    vkCmdPipelineBarrier(
        srcStageMask=FRAGMENT_SHADER,
        dstStageMask=FRAGMENT_SHADER,
        VkBufferMemoryBarrier { srcAccessMask=0, dstAccessMask=SHADER_READ }
    );

since the memory dependency will vanish if you swap the barriers.)

**Comment by krOoze** [comment]
how so? why would it vanish?

**Comment by philiptaylor** [comment]
Memory dependencies are defined to exist from (srcStageMask, srcAccessMask) to (dstStageMask, dstAccessMask) where both halves are defined either in the same barrier command or in two separate commands with an execution dependency chain between them. There's an execution dependency chain from the first `vkCmdPipelineBarrier` to the second, but not from the second to the first, so the memory dependency will only exist when the srcAccessMask half is in the first barrier.

**Comment by krOoze** [comment]
Sorry, deleted that follow-up post before you posted, to think more about it.

Right, the reordering thing only applies to Action commands. Though, I still don't see the difference in the example (other that the swapped version is less intuitive -- dependency chain is between pipeline stages and, not between barriers themselves).

Anyway the point is the ImageBarrier have the layout transition side-effect (which makes it part Action Command). So possibly they felt the need to specifically state the order (e.g. that two barriers do not write an Image memory concurently or in wrong order).

Okay, let's see what the authors have to say on the topic...

**Comment by jeffbolznv** [comment]
> I have no idea what this means. What was intended here?

It means you don't have to put a barrier between your barriers :-).

The spec doesn't say that a layout transition executes in any particular pipeline stage so the app has no way to order them against each other using execution dependencies (unless we required using ALL_COMMANDS_BIT, which would be expensive). IIRC we considered adding a pseudo-stage for pipeline transitions, but chose to say they are automatically ordered instead.

**Comment by philiptaylor** [comment]
Hmm... I suppose I interpret the execution dependency chain definition as giving a meaningful partial order between `vkCmdPipelineBarrier`s (where one barrier's srcStageMask overlaps the previous barrier's dstStageMask, etc), and memory dependencies sort of define the order of make-available and make-visible operations based on the execution dependency chain order (where make-available must come before the corresponding make-visible along a chain), and this bit:

> When an image memory barrier includes a layout transition, the barrier first makes writes via srcStageMask and srcAccessMask available, then performs the layout transition, then makes the contents of the image subresource(s) in the new layout visible to memory accesses in dstStageMask and dstAccessMask, as if there is an execution and memory dependency between the source masks and the transition, as well as between the transition and the destination masks.

sort of defines the order that transitions execute in relative to make-available and make-visible operations and other commands, and therefore it determines the order of transitions relative to each other based on the execution dependency chains. I think that (partial) order between transitions should be sufficient for sensible transition/barrier usage.

(I suppose my interpretation involves quite a lot of inferring/guessing from the spec, and people might not agree with it... I like to view everything in terms of ordered operations (where a pipeline barrier means "wait for previous draws/barriers to signal a particular set of stages, then perform a make-available operation (flush caches etc), then perform a transition, then make-visible, then signal a particular set of stages" (or any observably equivalent implementation)), because I think that makes the most sense, whereas the spec is a bit more abstract and declarative (and ambiguous and confusing, I think).)

Anyway, I'm still not really sure whether the sentence about being "automatically ordered" is just noting that there is already this order defined (loosely) by execution dependency chains - I don't really see what's "automatic" about that - or whether it's trying to declare some new ordering constraints that are (automatically) applied independent of execution dependencies (e.g. transitions will always be performed in the order they were submitted to the queue, or something).

---

If it helps to be a bit more concrete: By my current interpretation, this is okay:

    vkCmdPipelineBarrier(
        srcStageMask=TOP_OF_PIPE,
        dstStageMask=COLOR_ATTACHMENT_OUTPUT,
        VkImageMemoryBarrier { image=img0, oldLayout=PRESENT_SRC, newLayout=GENERAL }
    );
    vkCmdPipelineBarrier(
        srcStageMask=COLOR_ATTACHMENT_OUTPUT,
        dstStageMask=BOTTOM_OF_PIPE,
        VkImageMemoryBarrier { image=img0, oldLayout=GENERAL, newLayout=PRESENT_SRC }
    );

because there's an execution-dependency-chain-like order between the two barriers (via `COLOR_ATTACHMENT_OUTPUT`). But this is not okay:

    vkCmdPipelineBarrier(
        srcStageMask=TOP_OF_PIPE,
        dstStageMask=BOTTOM_OF_PIPE,
        VkImageMemoryBarrier { image=img0, oldLayout=PRESENT_SRC, newLayout=GENERAL }
    );
    vkCmdPipelineBarrier(
        srcStageMask=TOP_OF_PIPE,
        dstStageMask=BOTTOM_OF_PIPE,
        VkImageMemoryBarrier { image=img0, oldLayout=GENERAL, newLayout=PRESENT_SRC }
    );

because there's no execution dependency order and so the transitions could be performed out of order or overlapped (corrupting the image and causing undefined behaviour). I'd like to be more confident in the correctness of my intepretation.

**Comment by jeffbolznv** [comment]
I'm pretty sure our intent was that the second option should work, and that's what this sentence intends to guarantee. This way apps don't have to create phony dependency chains via random pipeline stages like in the first example, just to separate the layout transitions.

**Comment by philiptaylor** [comment]
If that's the intent, I think it needs to be specified in far more detail. Which other transitions are they ordered against (presumably not ones in other queues)? What order is used - presumably the order of command buffer submission, then the order of recording into command buffers (treating secondary buffers appropriately), then the order of multiple image barriers inside each pipeline barrier command; but where do subpass-dependency transitions fit into that? Do subpass-dependency transitions get automatically ordered against each other too? What happens when an implementation is allowed to omit a subpass-dependency transition because it treats the two layouts identically? Can the ordering between transitions function as a link in an execution dependency chain (if not then it seems hard to use memory dependencies correctly around automatically-ordered transitions)? etc.

(I don't really see how that kind of transition ordering would be helpful, though, and it feels weird to me compared to the rest of the API (which lets everything happen out-of-order by default). The first example is not really meant to be a phony dependency chain - it's what you might get if the application starts a frame, draws N objects onto the swapchain image, then presents the frame, in the degenerate case when N=0. I don't see why anyone would want to have multiple transitions of the same image without anything happening in the middle, since you might as well just merge them into a single transition, except as the degenerate case where you normally have something in the middle and already have the right execution dependencies between them.)

**Comment by krOoze** [comment]
@philiptaylor BTW Is there really any execution dependency even in the first example? It must rely on automatic ordering too. (It would only work as intended if draw command was between the two barriers)

Barriers are not **action** commands to which (other) barriers apply (well,maybe the ImageBarrier is due to the layout transition, and that is explained reasonably). But even if they were, they don't use any pipeline stages, so you won't be able to specify any execution dependency explicitly between them.

@jeffbolznv Is there any good uses of this? I was trying to think of any useful overlaping barriers situation. but came blank (they would just lead to undefined behavior). They should always rely on execution dependency through some other command.

Only valid use I think of is, when it just overrides the previous barrier (through the automatic ordering), it seems useless (and should probably be caught by layers as performance warning -- or removed to allow simpler drivers). Is my understanding correct?

**Comment by ratchetfreak** [comment]
@krOoze in 6.4 the execution dependence chain states that:

> for each dependency in the sequence (except the first) at least one of the following conditions is true:
>
> * srcStageMask of the current dependency includes at least one bit CC that is present in the dstStageMask of the previous dependency. Or [...]

So barriers can have execution dependencies between them.

**Comment by philiptaylor** [comment]
> BTW Is there really any execution dependency even in the first example?

By a literal interpretation of the spec, maybe not, or maybe (it's unclear). But I think the current spec text is not coherent enough to interpret literally, so I'm more interested in understanding the intended observable behaviour, and trying to derive some coherent model from that (in order to attempt to implement validation for it).

**Comment by TobiasHector** [comment]
@philiptaylor 

> By a literal interpretation of the spec, maybe not, or maybe (it's unclear). But I think the current spec text is not coherent enough to interpret literally, so I'm more interested in understanding the intended observable behaviour, and trying to derive some coherent model from that (in order to attempt to implement validation for it).

They are supposed to be able to depend on each other. It's roughly the same way as how semaphore signal/wait operations are queue operations, and thus can be depended on by other queue operations - though it's complicated by the pipeline stages. The execution chain section tries to explain this (in a really confusing manner), but that's the intent.

So what @ratchetfreak says is correct - they *can* have dependencies between them, but they don't *need* to have dependencies between them. This lets you define a DAG for dependencies, rather than just a straight line - allowing implementations to hide latency by doing other useful work if/when a stall occurs.

**Comment by krOoze** [comment]
@ratchetfreak You omit the context:

ch 2.2.1. Queue Operation:

> When not enforced by an explicit dependency or otherwise forbidden by the specification, action [!] commands may overlap execution or execute out of order, and may not see the side effects of each other’s memory accesses.

Same is not said for synchronization (or state) commands.

ch 6.4. Execution And Memory Dependencies:
> Synchronization commands introduce explicit execution and memory dependencies between two sets of action [!] commands, [...]

Is it an action command in the first place? Only partly action command is the ImageBarrier (with layout transition) and that dependency is (IMHO well enough) explained later in the chapter.

and

> An execution dependency chain from a set of source pipeline stages [!] A to a set of destination pipeline stages [!] B is a sequence of execution dependencies [...]

So through what pipeline stages the barrier passes exactly in its execution (IMHO none - only adds additional dependency to the cumulative set)?

@philiptaylor Yea, it probably falls under the umbrella of issue #132.

**Comment by TobiasHector** [comment]
@krOoze 
So I don't dislike the principle of describing action vs. synchronization vs. state commands - but the problem is we don't really describe what each command is amongst those categories. It also as written currently implies a mutual exclusion - which isn't really true. In fact all (most?) synchronization commands are practically ALSO action commands - something is always done as a result of executing a barrier or event (caches are flushed or invalidated, workloads are triggered, etc.), and these can be ordered against each other in roughly the same way as between a regular action command and a synchronization command. (NB: This is probably an oversimplification - but the principle should be correct)

**Comment by krOoze** [comment]
> [...] we don't really describe what each command is amongst those categories.

You dont? ch. 2.2.1. Queue Operation:
> Commands recorded in command buffers either perform actions (draw, dispatch, clear, copy, query/timestamp operations, begin/end subpass operations), set state (bind pipelines, descriptor sets, and buffers, set dynamic state, push constants, set render pass/subpass state), or perform synchronization (set/wait events, pipeline barrier, render pass/subpass dependencies).

<br/>
> synchronization commands are practically ALSO action commands

Yea, but you talk (transparent) implementation? What it is to me, if I can't catch them by any of the offered pipeline stages.

It comes as a surprise to me, I thought barriers are immediate (similar as state setting), i.e. just cumulating dependency descriptions -- not being subject for dependency themselves (except layout transition of course).

So you are saying these barriers will block themselves?:

    copy( from=img1, to=img2 );
    draw( FB=img3 );
    imgBarr( COLOR_ATT, TRANSFER, img3 ); // do not want the next barrier to force me to execute prematurely. I can wait until the copy.
    imgBarr( TRANSFER, FRAGMENT_SHADER, img2 ); // do not really want exe dependency on the previous barrier, but on the copy. I can execute way earlier.
    draw( texture=img2, FB=img5 )
    copy( from=img3, to=img4 );

**Comment by TobiasHector** [comment]
> You dont? ch. 2.2.1. Queue Operation:

It's a rough list that only vaguely calls out functions (and not by their proper names), and it's not exhaustive - particularly not in the face of extensions. Other command properties are called out explicitly on each function definition (e.g. queue capabilities, in- or outside of a render pass) - that's what I'm referring to.

> So you are saying these barriers will block themselves?:

> ```
copy( from=img1, to=img2 );
draw( FB=img3 );
imgBarr( COLOR_ATT, TRANSFER, img3 ); // do not want the next barrier to force me to execute prematurely. I can wait until the copy.
imgBarr( TRANSFER, FRAGMENT_SHADER, img2 ); // do not really want exe dependency on the previous barrier, but on the copy. I can execute way earlier.
draw( texture=img2, FB=img5 )
copy( from=img3, to=img4 );
```

When you say "block themselves", if you mean causing a cyclic dependency - then no, that's impossible.

If you mean that the latter waits on the former... then it's possible. My mind isn't in the right place atm to figure out if your example is definitely doing that (I *think* it is), but it's certainly possible to do it - that's basically what the execution dependency chains section is talking about.

**Comment by TobiasHector** [comment]
Separating this comment out because it's kind of important:

> Yea, but you talk (transparent) implementation? What it is to me, if I can't catch them by any of the offered pipeline stages.

It's not transparent, it's completely part of how the spec operates. The operation of a barrier has finished by the time the pipeline stages specified by the dstStageMask have started. So if you wait on those stages in a subsequent pipeline, that original barrier has completed (including the subsequent image layout transitions, queue ownership transfers, flushes/invalidates etc.). Again, this is basically what the "execution dependency chain" section is calling out.

**Comment by philiptaylor** [comment]
> So you are saying these barriers are dependent between themselves?:
> 
>     copy( from=img1, to=img2 );
>     draw( FB=img3 );
>     imgBarr( COLOR_ATT, TRANSFER, img3, appropriate mem dep. and layout switch ); // do not want the next barrier to force me to execute prematurely. I can wait until the copy.
>     imgBarr( TRANSFER, FRAGMENT_SHADER, img2, appropriate mem dep. and layout switch ); // do not really want exe dependency on the previous barrier, but on the copy. I can execute way earlier.
>     draw( texture=img2, FB=img5 )
>     copy( from=img3, to=img4 );

Yes, I believe the second barrier (and associated transition etc) will have to wait for the first barrier.

If there isn't any automatic order between transitions (beyond the normal execution dependency order), you could write this as:

    copy( from=img1, to=img2 );
    draw( FB=img3 );
    setEvent( ev1, COLOR_ATT );

    imgBarr( TRANSFER, FRAGMENT_SHADER, img2, appropriate mem dep. and layout switch );
    draw( texture=img2, FB=img5 );

    waitEvents( ev1, COLOR_ATT, TRANSFER, img3, appropriate mem dep. and layout switch );
    copy( from=img3, to=img4 );

Then (if I got it right) there should be no unwanted dependencies - every operation can start immediately after its inputs are available, and the transitions could be reordered or could run concurrently.

But if (per the original question in this issue) there is some vaguely-specified total order between transitions, you couldn't do that and the transitions would run serially.

**Comment by TobiasHector** [comment]
> But if (per the original question in this issue) there is some vaguely-specified total order between transitions, you couldn't do that and the transitions would run serially.

I *thought* that the only implicit ordering here was that image transitions happen between (the wait & any memory being made available), and (memory being made visible & continuing). Spec seems to disagree though, unless I'm missing something...

**Comment by TobiasHector** [comment]
Yea I'm not missing anything, we really did do that, and I do vaguely recall the conversation @jeffbolznv mentioned. So, there is an order between transitions, though crucially this does NOT guarantee ordering between the WHOLE barrier - the barrier itself is not guaranteed to complete just because its transitions have.

**Comment by oddhack** [comment]
This should be fixed in the 1.0.35 spec update, as part of the synchronization language rewrite.
